README.md
# WFST for Ukrainian ITN

Simple WFST for Ukrainian ITN based on NVIDIA NeMo and Pynini

## Installation

```shell
pip install ukr-itn
```

## Usage

```python
from ukr.wfst import normalize

normalize("це трапилося дві тисячі дев'ятнадцятого числа")  # це трапилося 2019 числа
normalize("мінус пять цілих одна десята відсотка")  # -5.1 %
normalize("двадцять дві тисячі сто один")  # 22101
```

### From command line

```shell
echo "це трапилося дві тисячі дев'ятнадцятого числа" | python -m ukr
```

```
Options:
  -h, --help     Show this help message and exit
  -j, --json     Return result as JSON
  -v, --verbose  Print original input and normalized to compare

```

Will return `це трапилося 2019-го числа`

### JSON output

For more advanced usage you can get json output

```python
from ukr.wfst import normalize

normalize("це трапилося дві тисячі дев'ятнадцятого числа", json=True)
# >>> '[{"word": "це"}, {"word": "трапилося"}, {"ordinal": "2019"}, {"word": "числа"}]' 
```

## How it works

We have two king of FST: taggers and verbalizers

This is a tagger:

```python
from ukr.wfst import classifyFst, apply_fst_text

apply_fst_text("мінус пять цілих одна десята відсотка", classifyFst.fst)  
```

will return `"measure { decimal { negative: "true" integer_part: "5" fractional_part: "1" } units: "%" }"`

And this is a verbalizers

```python
from ukr.wfst import verbalizeFinalFst, apply_fst_text

apply_fst_text('measure { decimal { negative: "true" integer_part: "5" fractional_part: "1" } units: "%" }', verbalizeFinalFst.fst)  
```

will return `-5.1 %`

CHANGELOG.md
# Changelog

### v0.1.8

- Added TIME class, some examples:
    - `"сьома година двадцять п'ять хвилин"` -> `'07:25'`
    - `"о пів на десяту"` -> `'09:30'`
    - `"пів на третю"` -> `'02:30'`
    - `"чверть на одинадцяту"` -> `'10:15'`
    - `"за чверть одинадцята"` -> `'10:45'`
    - `"п'ять хвилин на дванадцяту"` -> `'11:05'`
    - `"дванадцята нуль нуль"` -> `'12:00'`
    - `"одинадцята нуль шість"` -> `'11:06'`
    - `"шоста сорок три"` -> `'06:43'`
- Defined new method `normalize` which should be used istead of `apply_fst_text` in production code.
  The reason is we need additional code to process some non-deterministic cases like for time class (for example minutes goes before hours).

### v0.1.7

- Added DATE class:
    - `"першого січня дві тисячі першого року"` -> `1 січня 2001 року'`
    - `"першого січня"` -> `1 січня'`
    - `"січень дві тисячі першого рок у"` -> `січень 2001 року'`
    - `"січень дві тисячі першого"` -> `січень 2001'`
    - `"дві тисячі першого рік"` -> `2001 рік'`
    - `"дев'ятсот сорок п'ятий рік до нашої ери"` -> `945 рік до н. е.'`
- Added JSON option for command line

### v0.1.6

- Added command line execution mode: `echo "це трапилося дві тисячі дев'ятнадцятого числа" | python -m ukr`
- Fixed bug with single digits, like `"один одного"` (but was `"1 1"`).
  Please note: single digit will normalize as written, like one -> one, two - two, ..., but eleven -> 11, twelve -> 12

ukr/taggers/time.py
import pynini

from ukr.graph_utils import (
    GraphFst,
    delete_extra_space,
    delete_space,
    NEMO_DIGIT,
    NEMO_CHAR,
)
from pynini.lib import pynutil

from ukr.taggers.cardinal import CardinalFst
from ukr.taggers.ordinal import OrdinalFst
from ukr.utils import get_abs_path


class TimeFst(GraphFst):
    """
    Finite state transducer for classifying time
        e.g. сьома година п'ять хвилин" -> time { hours: "07" minutes: "05" }
        e.g. о пів на десяту -> time { hours: "09" minutes: "30" }
        e.g. чверть на десяту -> time { hours: "09" minutes: "15" }
        e.g. за чверть одинадцята -> time { hours: "10" minutes: "45" }
        e.g. п'ять хвилин на дванадцяту -> time { hours: "11" minutes: "05" }

        e.g. twelve past one -> time { minutes: "12" hours: "1" }
        e.g. two o clock a m -> time { hours: "2" suffix: "a.m." }
        e.g. quarter past two -> time { hours: "2" minutes: "15" }
    """

    def __init__(self, cardinal: CardinalFst, ordinal: OrdinalFst):
        super().__init__(name="time", kind="classify")

        hours = pynini.string_file(get_abs_path("data/time/hours.tsv"))
        minutes = pynini.string_file(get_abs_path("data/time/minutes.tsv"))
        to_hour = pynini.string_file(get_abs_path("data/time/to_hour.tsv"))
        zeros = cardinal.graph_zero

        hours_ordinal = ordinal.graph_up_to_hundred_component
        hours_ordinal = hours_ordinal @ (pynini.closure(NEMO_DIGIT) + pynutil.delete(pynini.union("-") + pynini.closure(NEMO_CHAR)))

        graph_hours = hours_ordinal + delete_space + pynini.closure(pynutil.delete(hours), 0, 1)
        graph_hours = pynutil.insert("hours: \"") + graph_hours + pynutil.insert("\"")

        minutes_cardinal = cardinal.graph_up_to_hundred_component + delete_space + pynini.closure(pynutil.delete(minutes), 0, 1)
        graph_minutes = pynutil.insert("minutes: \"") + minutes_cardinal + pynutil.insert("\"")

        graph_half_hour = pynini.closure(pynutil.delete("о "), 0, 1) + pynutil.delete("пів на ") + hours_ordinal
        graph_half_hour = graph_half_hour @ to_hour.invert()
        graph_half_hour = pynutil.insert("hours: \"") + graph_half_hour + pynutil.insert("\" minutes: \"30\"")

        to_hour_graph = hours_ordinal @ to_hour
        graph_to_quarter_hour = pynutil.delete("чверть на ") + to_hour_graph
        graph_to_quarter_hour = pynutil.insert("hours: \"") + graph_to_quarter_hour + pynutil.insert("\" minutes: \"15\"")

        graph_from_quarter_hour = pynutil.delete("за чверть ") + to_hour_graph
        graph_from_quarter_hour = pynutil.insert("hours: \"") + graph_from_quarter_hour + pynutil.insert("\" minutes: \"45\"")

        graph_hm = graph_hours + delete_extra_space + graph_minutes

        # NOTE: we use here special notation >> which will be processed after FST
        # The >> means move the current token to one position to right.

        # п'ять хвилин на дванадцяту -> time { hours: "11" minutes: "05" }
        graph_mh = graph_minutes + pynutil.insert(">>") + pynutil.delete(" на ") + pynutil.insert(" hours: \"") + to_hour_graph + pynutil.insert("\"")

        # дванадцята нуль нуль ->  time { hours: "12" minutes: "00" }
        graph_hzz = graph_hours + delete_space
        graph_hzz += pynini.union(
            pynutil.insert(" minutes: \"") + zeros + delete_space + zeros + pynutil.insert("\""),
            pynutil.insert(" minutes: \"") + zeros + delete_space + cardinal.graph_digit + pynutil.insert("\""),
        )

        final_graph = graph_hm | graph_mh | graph_half_hour | graph_to_quarter_hour | graph_from_quarter_hour | graph_hzz
        final_graph = self.add_tokens(final_graph.optimize())

        self.fst = final_graph.optimize()
